### PR TITLE
Add support for Pub versioning

### LINKED ISSUE #708: Add support for Pub versioning
Pub closely follows semver, but unfortunately not exactly: https://pub.dev/packages/pub_semver

### PR DESCRIPTION
**Pull request by michaelkedar**
Closes #708

### CODE REVIEW

File: osv/semver_index.py [hunk 1]
```diff
@@ -42,14 +42,26 @@ def _remove_leading_zero(component):
 def coerce(version):
   """Coerce a potentially invalid semver into valid semver."""
   version = _strip_leading_v(version)
-  version_pattern = re.compile(r'^(\d+)(\.\d+)?(\.\d+)?(.*)$')
+  version_pattern = re.compile(r'^(\d+)(\.\d+)?(\.\d+)?(([^+]*)(.*))$')
   match = version_pattern.match(version)
   if not match:
     return version
 
+  # also remove leading 0's from pre-release suffix
```
Comment (oliverchang) [inline osv/semver_index.py:50]: Can you add a comment that this is to allow versions allow from earlier Semver 2.0.0 specs to be parsed as well?

File: osv/pub.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# https://semver.org/spec/v2.0.0-rc.1.html
+# Differences from SemVer 2.0.0 are described at
+# https://pub.dev/packages/pub_semver
+# Only difference that affects us is the inclusion of build suffix when
+# ordering.
+#  - Build suffixes are parsed the same as pre-release suffixes.
+#  - Pre-release suffixes are evaluated before build suffixes.
+#  - Versions with build suffixes come after versions without.
+#    e.g. 1.0.0-pre < 1.0.0-pre+build < 1.0.0 < 1.0.0+build
+# SemVer 2.0.0-rc.1 also does not explcitly disallow empty identifiers or
```
Comment (oliverchang) [inline osv/pub.py:30]: Wow, nice catch! How did you notice this case?

File: osv/semver_index.py [hunk 3]
```diff
@@ -42,14 +42,26 @@ def _remove_leading_zero(component):
...
+  if match.group(4) and match.group(4)[0] == '-':
+    pre_components = []
```
Comment (oliverchang) [inline osv/semver_index.py:52]: Can you move all of this into a `_remove_leading_zeros` function?

File: osv/pub.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+@functools.total_ordering
+class Version:
+  """Pub version."""
+
+  def __init__(self, semver):
+    # Tuple comparison of (semver,) or (semver, build) is consistent with
+    # desired ordering.
+    if semver.build:
+      self._version = (semver, semver_index.normalize_prerelease(semver.build))
```
Comment (oliverchang) [inline osv/pub.py:43]: Can you add a comment that we re-use semver_index.normalize_prerelease to enable lexicographic sort?

**Review by oliverchang — COMMENTED** [review]
very nice! 

Let's also make that refactor you mentioned as well in a subsequent PR.

File: osv/ecosystems.py [hunk 5]
```diff
@@ -657,6 +658,40 @@ def enumerate_versions(self,
                                        last_affected, limits)
 
 
+class Pub(Ecosystem):
+  """Pub ecosystem"""
+
+  _API_PACKAGE_URL = 'https://pub.dev/api/packages/{package}'
```
[the remark this answers is not in the document]
Comment (michaelkedar) [reply osv/ecosystems.py:664]: There's no mention of a rate limit in https://pub.dev/help/api

File: osv/pub.py (hunk 2 above)
Comment (michaelkedar) [reply osv/pub.py:30]: I was comparing 2.0.0 and 2.0.0-rc.1 side by side, and noticed the explicit callout in 2.0.0 `Identifiers MUST NOT be empty.` was not in rc.1 🙂

I noted this, but then did not actually check how our system reacts to it. The [tests in pub_semver](https://github.com/dart-lang/pub_semver/blob/master/test/version_test.dart) don't seem to test for it either.

I don't even know if Pub actually allows it, but I'll add in a test in anyway.

File: osv/pub.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+@functools.total_ordering
+class Version:
+  """Pub version."""
+
+  def __init__(self, semver):
+    # Tuple comparison of (semver,) or (semver, build) is consistent with
+    # desired ordering.
+    if semver.build:
+      # Reuse normalize_prerelease to so that the
+      # build suffix can to be compared lexicographically.
```
Comment (oliverchang) [inline osv/pub.py:44]: s/can to be/can be/

**Review by oliverchang — APPROVED** [review]
LGTM with a tiny nit. Thanks!

**Review by oliverchang — APPROVED** [review]

**Pull request merged**